cidc_api/shared/gcloud_client.py

```diff
@@ -121,60 +122,174 @@ def revoke_upload_access(user_email: str):
     bucket.set_iam_policy(policy)
 
 
-def grant_download_access(user_email: str, trial_id: str, upload_type: str):
+def _build_intake_prefix(
+    user_id: int, user_email: str, trial_id: str, upload_type: str
+) -> str:
+    usernameish = user_email.split("@")[0].replace(".", "")
+    return (
+        f"{_build_trial_upload_prefix(trial_id, upload_type)}/{usernameish}-{user_id}"
+    )
+
+
+def grant_intake_access(
+    user_id: int, user_email: str, trial_id: str, upload_type: str
+) -> str:
+    """
+    Give a user upload access to a subdirectory of the intake bucket with prefix like
+    <trial id>/<upload type>/<user email prefix>-<user id>
+
+    Return the GCS URI to which access has been granted.
+    """
+    prefix = _build_intake_prefix(user_id, user_email, trial_id, upload_type)
+
+    logger.info(f"Granting intake access on {prefix} to {user_email}")
+
+    return grant_conditional_gcs_access(
+        GOOGLE_INTAKE_BUCKET, prefix, GOOGLE_UPLOAD_ROLE, user_email
+    )
+
+
+def revoke_intake_access(
+    user_id: int, user_email: str, trial_id: str, upload_type: str
+) -> str:
+    """
+    Revoke a user's upload access to a subdirectory of the intake bucket with prefix like
+    <trial id>/<upload type>/<user email prefix>-<user id>
+
+    Return the GCS URI from which access has been revoked.
+    """
+    prefix = _build_intake_prefix(user_id, user_email, trial_id, upload_type)
+
+    logger.info(f"Granting intake access on {prefix} to {user_email}")
+
+    return revoke_conditional_gcs_access(
+        GOOGLE_INTAKE_BUCKET, prefix, GOOGLE_UPLOAD_ROLE, user_email
+    )
+
+
+user_member = lambda email: f"user:{email}"
+
+intake_subdir_regex = re.compile(
+    f'resource.name.startsWith\("projects/_/buckets/{GOOGLE_INTAKE_BUCKET}/objects/(.*?)"\)'
+)
+
+
+def list_intake_access(user_email: str) -> List[str]:
+    """
+    List the GCS URIs in the intake bucket to which this user has access.
+    """
+    bucket = _get_bucket(GOOGLE_INTAKE_BUCKET)
+    policy = bucket.get_iam_policy(requested_policy_version=3)
+    policy.version = 3
+
+    user_uris = []
+    for binding in policy.bindings:
+        expression = binding.get("condition", {}).get("expression", "")
+        subdir_match = intake_subdir_regex.match(expression)
+        if (
+            binding["members"] == {user_member(user_email)}
```

Comment:
should this be `user_member(user_email) in binding["members"]`?

Reply:
So, we're actually looking specifically for bindings that contain *only* this user as a member - if that weren't the case, then `in` would be more appropriate. I'll add a comment to this effect.